packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -110,14 +171,111 @@ class WebSocketConnection implements Closeable {
   @override
   void close([int closeStatus = _defaultCloseStatus]) {
     _logger.verbose('Closing web socket connection.');
+    _hubEventsController?.add(SubscriptionHubEvent.disconnected());
+    _rebroadcastController.close();
+
+    _resetConnectionInit();
+    _resetConnectionVariables(closeStatus);
+
+    _subscriptionRequests.clear();
+    _hubEventsController?.close();
+    _hubEventsController = null;
+  }
+
+  /// Initiate polling
+  void _startPolling() {
+    final interval = _subscriptionOptions?.pingInterval ?? _defaultPingInterval;
+
+    _pingTimer?.cancel();
+    _pingTimer = Timer.periodic(interval, (_) => _poll());
+    _poll();
+  }
+
+  /// Pings AppSync to verify health of connection. Failed pings trigger
+  /// reconnection through retry/back off.
+  Future<void> _poll() async {
+    try {
+      if (_pingClient == null) return;
+      final res = await _getPollRequest();
+      if (res!.body != 'healthy') {
+        throw Exception([
+          'Subscription connection broken.',
+          'AppSync status check returned: ',
+          res.body
+        ]);
+      }
+    } on Exception catch (e) {
+      _pingTimer?.cancel();
+      _timeoutTimer?.cancel();
+
+      // flag for reconnection when network returns
+      _hasConnectionBroken = true;
+
+      // only try to recover if network is turned on
+      if (_hasNetwork) {
+        _retry();
+      } else {
+        _hubEventsController?.add(SubscriptionHubEvent.connecting());
```

Comment:
I would recommend only using `?` when we really don't care if it's null. Here I think we absolutely need it to be non-null - otherwise this event is never reported, so I would use `!` and ensure that's always true with tests.

Same throughout as well

Reply:
_huEventsController could really be a `final` var on the class since there's no real reason to close it ever